#! /bin/sh /usr/share/dpatch/dpatch-run
## 38_hurd.dpatch by  <glandium@debian.org>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Fix FTBFS on Hurd because of MAXPATHLEN

@DPATCH@

--- nss/mozilla/security/nss/cmd/shlibsign/shlibsign.c.orig
+++ nss/mozilla/security/nss/cmd/shlibsign/shlibsign.c
@@ -164,7 +164,6 @@
 #ifdef USES_LINKS
     int ret;
     struct stat stat_buf;
-    char link_buf[MAXPATHLEN+1];
     char *link_file = NULL;
 #endif
 
@@ -271,10 +270,22 @@
     }
     if (S_ISLNK(stat_buf.st_mode)) {
 	char *dirpath,*dirend;
-	ret = readlink(input_file, link_buf, sizeof(link_buf) - 1);
-	if (ret < 0) {
-	   perror(input_file);
-	   goto loser;
+	char *link_buf = NULL;
+	size_t size = 64;
+	while (1) {
+	    link_buf = realloc(link_buf, size);
+	    if (!link_buf) {
+	       perror(input_file);
+	       goto loser;
+	    }
+	    ret = readlink(input_file, link_buf, size - 1);
+	    if (ret < 0) {
+	       perror(input_file);
+	       goto loser;
+	    }
+	    if (ret < size - 1)
+	    	break;
+	    size *= 2;
 	}
 	link_buf[ret] = 0;
 	link_file = mkoutput(input_file);
